Library/Homebrew/cmd/reinstall.rb

```diff
@@ -15,172 +15,174 @@
+          }],
+        ].each do |args|
+          options = args.pop
+          send(*args, **options)
+          conflicts "--cask", args.last
+        end
+        formula_options
+        [
+          [:switch, "--cask", "--casks", { description: "Treat all named arguments as casks." }],
+          [:switch, "--[no-]binaries", {
+            description: "Disable/enable linking of helper executables (default: enabled).",
+            env:         :cask_opts_binaries,
+          }],
+          [:switch, "--require-sha",  {
+            description: "Require all casks to have a checksum.",
+            env:         :cask_opts_require_sha,
+          }],
+          [:switch, "--[no-]quarantine", {
+            description: "Disable/enable quarantining of downloads (default: enabled).",
+            env:         :cask_opts_quarantine,
+          }],
+          [:switch, "--adopt", {
+            description: "Adopt existing artifacts in the destination that are identical to those being installed. " \
+                         "Cannot be combined with `--force`.",
+          }],
+          [:switch, "--skip-cask-deps", {
+            description: "Skip installing cask dependencies.",
+          }],
+          [:switch, "--zap", {
+            description: "For use with `brew reinstall --cask`. Remove all files associated with a cask. " \
+                         "*May remove files which are shared between applications.*",
+          }],
+        ].each do |args|
+          options = args.pop
+          send(*args, **options)
+          conflicts "--formula", args.last
+        end
+        cask_options
+
+        conflicts "--build-from-source", "--force-bottle"
+
+        named_args [:formula, :cask], min: 1
       end
-    end
-
-    Install.perform_preinstall_checks
 
-    formulae.each do |formula|
-      if formula.pinned?
-        onoe "#{formula.full_name} is pinned. You must unpin it to reinstall."
-        next
+      sig { override.void }
+      def run
+        formulae, casks = T.cast(
+          args.named.to_formulae_and_casks(method: :resolve).partition { _1.is_a?(Formula) },
+          [T::Array[Formula], T::Array[Cask::Cask]],
+        )
+
+        if args.build_from_source?
+          unless DevelopmentTools.installed?
+            raise BuildFlagsError.new(["--build-from-source"], bottled: formulae.all?(&:bottled?))
+          end
+
+          unless Homebrew::EnvConfig.developer?
+            opoo "building from source is not supported!"
+            puts "You're on your own. Failures are expected so don't create any issues, please!"
+          end
+        end
+
+        Install.perform_preinstall_checks
+
+        formulae.each do |formula|
+          if formula.pinned?
+            onoe "#{formula.full_name} is pinned. You must unpin it to reinstall."
+            next
+          end
+          Migrator.migrate_if_needed(formula, force: args.force?)
+          Homebrew.reinstall_formula(
```

Comment:
This should probably get namespaced properly at some point. It seems to get used in both the upgrade and reinstall commands.